Simple improvements

pmanna opened the pull request:
- Made the generic sendRequestWithParameters: public, in order to deal with special requests (like session start)
- Added optional support for iOS 6
- Added simple persistent storage for requests when offline: requests are stored on disk when app becomes inactive and re-queued when connection is available again

Commit:
- Made the generic sendRequestWithParameters: public, in order to deal with special requests (like session start)
- Added optional support for iOS 6
- Added simple persistent storage for requests when offline

emiliopavia commented:
Hi @pmanna, thanks for your contribution! I've considered to add persistency of requests when offline, but I've decided to skip it because Google Analytics automatically drops requests that are older than 4 hours, and maybe that complexity wasn't worth. Anyway I'll look at your implementation.

pmanna commented:
Thanks for your reply!
To explain the rationale behind it: I work for a newspaper publisher, and page views are not just a statistics tool, but the foundations on which the advertising work to evaluate performance, so anything that could be done to minimise the losses of data MUST be done, and persistence is a requirement.
We're actually ditching Google's own SDK precisely because we've found out that it throttles the output, discarding events when the queue is full, and this is not acceptable.
Besides, the issue you report is easily solved: just use the date the event is sent, not when it's queued. I know, not exactly right, still it doesn't really matters as long as the total number of events is the right one.

However, I'm going to change the code again to ensure all the events are actually queued, not just the failed ones: let me know if that could be interesting for you, otherwise I'll keep them in my fork.

Thank you again for your work on this matter, it spared me a few hours of coding!

Commit:
All events are now queued & persisted, not only the failed ones, to limit throughput to GA servers and preventing drops at the same time.

Commit:
- Made iOS 6 support dynamic
- Shortened throttling interval

Commit:
Workaround for iOS 7 bug where identifierForVendor returns NULL

Commit:
Nasty bug: corrected parameter encoding, as values containing special characters (like '&') were NOT encoded properly by "stringByAddingPercentEscapesUsingEncoding:"

Commit:
- Added mutex control to prevent race conditions
- Added better emulation of GA SDK 3 parameters
- Added support for IDFA parameter

Commit:
Corrected edge cases and potential iOS 6&7 crashes when IDFA is nil

Commit:
Substituted NSTimer with GCD timer source
Repositioned mutex lock/unlock

Commit:
- Aligned REST calls to iOS SDK version 3.1.5
- Better "weak self" handling in blocks

Commit:
Added code to support hit validation while in DEBUG mode
https://developers.google.com/analytics/devguides/collection/protocol/v1/validating-hits
